File: robot.php
<?php

if(!empty($argv[1]) && file_exists($argv[1])){
	$f = fopen($argv[1], "r");
}

$robot = new Robot();
if(empty($f)){
	$f = fopen("php://stdin", "r");
	echo "----------------------\nPlease input following commands\n";
	echo "PLACE 0,0,NORTH\n";
	echo "LEFT\n";
	echo "RIGHT\n";
	echo "MOVE\n";
	echo "REPORT\n";
	echo "EXIT\n----------------------\n";
}
while ($line = fgets($f)){
	if(strstr($line,"EXIT")) exit;
	$line = trim($line);
	$array = array_filter(explode(" ",$line));
	
	if(empty($array)) continue;
	
	switch($array[0]){
		case "PLACE":
			$args = implode($array);
			$args = str_replace("PLACE",'',$args);
			$args = explode(",",$args);
			if(count($args)!=3) {
				echo "Invald Arguments.\n";	
				break;
			}
			$robot->PLACE($args[0],$args[1],$args[2]);	
			break;
		case "LEFT":
			$robot->LEFT();
			break;
		case "RIGHT":
			$robot->RIGHT();
			break;
		case "MOVE":
			$robot->MOVE();
			break;
		case "REPORT":
			$robot->REPORT();
			break;
		case "EXIT":
			exit;
			break;
		default:
			echo "Invalid Command.\n";
			break;
	}
}

class Status{
	const ACTIVE = '1';
	const INACTIVE = '0';
}

class Robot{
	private $x;
	private $y;
	private $f;
	private $status;
	private $FACE;
	private $firstPlacement;
	
	public function __construct(){
		$this->FACE = array(
			0 => "NORTH",
			1 => "EAST",
			2 => "SOUTH",
			3 => "WEST");
		$this->status = Status::INACTIVE;
		$this->firstPlacement = true;
	}
	
	public function PLACE($x,$y,$fString){
	
		
	
		if(!in_array($fString, $this->FACE)) {
			echo "Invald Argument.\n";	
			return;
		}
		
		if($this->firstPlacement && $this->_isOutOfRange($x,$y)){
			echo "Initial placement cannot out of table\N";
			return;
		}
		
		$this->x = $x;
		$this->y = $y;
		
		
		$this->f = array_search($fString, $this->FACE);
				
		if($this->_isOutOfRange($this->x,$this->y)){
			$this->status = Status::INACTIVE;

		}else{
			$this->status = Status::ACTIVE;               
		}
		
		$this->firstPlacement = false;
	}
		
	public function MOVE(){
		if($this->status == Status::INACTIVE){
			echo "WARNING:Command is ignored.\n";
			return;
		}
		$resultArray = $this->_tryToMove($this->x,$this->y,$this->f);
		if(empty($resultArray)){
			echo "WARNING:Robot cannot go further.\n";		
		}else{
			$this->x = $resultArray['x'];
			$this->y = $resultArray['y'];		
		}
	}
		
	public function LEFT(){
		if($this->status == Status::INACTIVE){
			echo "WARNING:Command is ignored.\n";
			return;
		}
		$this->f = $this->f - 1;
		if($this->f < 0){
			$this->f = $this->f + 4;
		}
		  
	}
	
	public function RIGHT(){
		if($this->status == Status::INACTIVE){
			echo "WARNING:Command is ignored.\n";
			return;
		}
		$this->f = $this->f + 1;
		if($this->f > 3){
			$this->f = $this->f - 4;
		}
		
	}
	
	public function REPORT(){
		if($this->status == Status::INACTIVE){
			echo "WARNING:Command is ignored.\n";
			return;
		}
		echo "Output: $this->x,$this->y,".$this->FACE[$this->f]."\n";
	}
	
	private function _isOutOfRange($x,$y){
		if($x>5 || $x<0 || $y>5 || $y<0){
			return true;
		}else{
			return false;
		}
	}
	
	private function _tryToMove($x,$y,$f){
		switch ($f) {
    		case 0:
        		$y++;
        		break;
    		case 1:
        		$x++;
        		break;
    		case 2:
    		    $y--;
     	   		break;
     	   	case 3:
    		    $x--;
     	   		break;
     	   	default:
     	   		break;
		}
		if($this->_isOutOfRange($x,$y))	{
			return array();
		}else{
			return array('x'=>$x,'y'=>$y);
		}	
	}
}

?>

File: robot.rb
#!/usr/bin/env ruby

class Robot
	@x
	@y
	@f
	@status
	@FACE
	@firstPlacement
	
	def initialize
		@status = :INACTIVE
		@FACE = Array["NORTH","EAST","SOUTH","WEST"]
		@firstPlacement = true
	end
	
	def place(x,y,fString)		
		if(@FACE.include?(fString))
			@f = @FACE.index(fString).to_i
		else
			puts "Invald Face Direction.\n";	
			return
		end
		
		
		if(@firstPlacement && isOutOfRange(x,y))
			puts "Initial placement cannot out of table\n"
			return
		end
		
		@x = x.to_i
		@y = y.to_i
		
		if(isOutOfRange(@x,@y))
			@status = :INACTIVE
		else
			@status = :ACTIVE
		end	
		
		@firstPlacement = false

	end
	
	def move
		if(@status == :INACTIVE)
			puts "WARNING:Command is ignored.\n"
			return
		end
		
		resultArray = tryToMove(@x,@y,@f)
		
		
		if(resultArray.count == 0)
			puts "WARNING:Robot cannot go further.\n"
		else
			@x = resultArray['x']
			@y = resultArray['y']		
		end
	end
	
	
	def left
		if(@status == :INACTIVE)
			puts "WARNING:Command is ignored.\n"
			return
		end
		
		@f = @f - 1;
		if(@f < 0)
			@f = @f + 4;
		end
	end
	
	def right
		if(@status == :INACTIVE)
			puts "WARNING:Command is ignored.\n"
			return
		end
		
		@f = @f + 1;
		if(@f > 3)
			@f = @f - 4;
		end
	end
	
	def report
		if(@status == :INACTIVE)
			puts "WARNING:Command is ignored.\n"
			return
		end
		puts "Output: #{@x},#{@y},#{@FACE[@f]}\n"
	end
	

	def processInput(line)
		if(line.index('EXIT'))
			exit!
		end
		
		line.strip!
		array = line.split(" ")
		
		if (array.count == 0) 
			return
		end
		
		command = array[0]
		
		case command
			when "PLACE"
				
				if (array.count == 2 )
					args = array[1].split(",")
					if (args.count == 3 )
						#do place
						place(args[0],args[1],args[2])
					end
				else
					puts "Invald Arguments.\n"
				end
			
			when "LEFT"
				left
				#do left
			when "RIGHT"
				right
				#do right
			when "MOVE"
				move
				#do move
			when "REPORT"
				report
				#do report
			when "EXIT"
			  exit!
			else
			  puts "Invalid Command.\n"
		end
		
	end
	
	
	private
	
	def isOutOfRange(x,y)
		x = x.to_i
		y = y.to_i
	
		if(x > 5 || x < 0 || y > 5 || y < 0)
			return true
		else
			return false
		end
	end
	
	def tryToMove(x,y,f)
			
		case f
    		when 0
				y+=1
    		when 1 
				x+=1
    		when 2 
				y-=1
     	   	when 3 
				x-=1 				
		end
	
		if(isOutOfRange(x,y))	
			return {}
		else
			return {'x'=>x,'y'=>y}
		end	
	end
	
end


r = Robot.new
if(ARGV.count > 0 && FileTest::exist?(ARGV[0]))
	
	File.open(ARGV[0], "r") do |infile|
		while (line = infile.gets)

			r.processInput(line)
			
        end
  	end
  	exit!
else

	puts "----------------------\nPlease input following commands\n";
	puts "PLACE 0,0,NORTH\n";
	puts "LEFT\n";
	puts "RIGHT\n";
	puts "MOVE\n";
	puts "REPORT\n";
	puts "EXIT\n----------------------\n";

	while line = gets
		r.processInput(line)
	end
	

end

